Add ROW_NUMBER(), RANK(), DENSE_RANK(), PERCENT_RANK(), and CUME_DIST() window functions

katzyn opened the pull request:
`ROW_NUMBER() OVER ([…])`, `RANK() OVER ([…])`, and `DENSE_RANK() OVER ([…])` are implemented. Only `PARTITION BY` and `ORDER BY` clauses are supported. Missing `ORDER BY` clause is accepted and returns results like in PostgreSQL (some databases forbid it). These functions may be used in grouped queries too, all necessary support was introduced for window aggregates earlier.

H2's `ROWNUM()` function is not touched and it still works as a plain non-window function. Before these changes `ROW_NUMBER() OVER ()` was incorrectly implemented as an alternative name for `ROWNUM()`, now they are entirely different.

Documentation is updated. Window functions and window specification for aggregates are marked as experimental.

@lukaseder

katzyn committed:
Add ROW_NUMBER(), RANK(), and DENSE_RANK() window functions

katzyn committed:
Prevent incorrect optimizations

katzyn committed:
Prevent incorrect optimizations for window aggregates too

lukaseder commented:
Excellent, thanks for pinging. I will try to build H2 this week and run the jOOQ integration tests against your implementation. I'll report any issues I might find.

katzyn committed:
Add a test

katzyn commented:
Issue #1161 is also fixed.

katzyn committed:
Fix window specification grammar in help.csv

katzyn committed:
Fix an unstable test

katzyn committed:
Add PERCENT_RANK() and CUME_DIST() window functions

katzyn commented:
I also added here implementations of `PERCENT_RANK()` and `CUME_DIST()` and fixed caching of results to avoid their evaluation for each row in partition.
